Initial breadcrumbs implementation.

bretthoerner opened the pull request:
Definitely still need to add tests.

Also, this is just the "backend" implementation right now. It all works but you have to build your own `Event`:

```java
        Raven raven = DefaultRavenFactory.ravenInstance();

        Breadcrumb b1 = new BreadcrumbBuilder()
            .setMessage("Breadcrumb 1")
            .setLevel("info")
            .setCategory("category")
            .build();
        Breadcrumb b2 = new BreadcrumbBuilder()
            .setMessage("Breadcrumb 2")
            .setLevel("warning")
            .setCategory("category")
            .build();

        List<Breadcrumb> breadcrumbs = new ArrayList<Breadcrumb>();
        breadcrumbs.add(b1);
        breadcrumbs.add(b2);

        Event event = new EventBuilder()
            .withMessage("Hello, world.")
            .withBreadcrumbs(breadcrumbs)
            .build();

        raven.sendEvent(event);
```
So my question now is... how are we handling these in other language loggers? It seems like they'd have to configure a second level (for capturing breadcrumbs) so that they could do something like "send error and above to Sentry, but capture warn and above as Breadcrumbs"?

dcramer commented:
// @mitsuhiko 

we talked about having a config which said "for this logger, only capture breadcrumbs of this level or higher", but the default is all things of INFO or higher (i think)

tkaemming commented on raven/src/main/java/com/getsentry/raven/event/Breadcrumb.java:
```diff
@@ -0,0 +1,111 @@
+    /**
+     * (Optional) Duration of the portion this breadcrumb represents.
+     */
+    private final float duration;
```
I'm not really sure how to interpret this commit, but I think this may have been removed here: getsentry/sentry@fa7a4159ca778d7516f01a7a74e30d33ff0a8fa5

tkaemming commented:
> So my question now is... how are we handling these in other language loggers? It seems like they'd have to configure a second level (for capturing breadcrumbs) so that they could do something like "send error and above to Sentry, but capture warn and above as Breadcrumbs"?

The Python implementation patches around the `logging` module to add the hooks. I think it'd make sense to add the ability to set up a second appender for breadcrumbs, and afterwards see if we can make it automatically installed when the `Raven` class is instantiated with whatever logging framework is available (with the ability to opt-out if you don't want them, or manually configure things like filters.)

bretthoerner commented:
OK, I've made a handful of changes,

* Added `RavenContext` which is more or less a port of the code from Python (thread local variable holds a set of `RavenContext` objects, which in turn hold `Breadcrumb`s -- may hold more in the future).
* When `Raven` is constructed, a `RavenContext` object is created and stored on it -- like Python.
* `Breadcrumbs` helper class added, which lets `record` be used to add a breadcrumb to all activate contexts -- like Python.
* `CircularFifoQueue` added from Apache Collections to hold last N `Breadcrumb`s on `RavenContext`.
* Added `ContextBuilderHelper`, which is used by `DefaultRavenFactory` and pulls any existing Breadcrumbs out of the `Raven` object's context.

There's still work to be done here, because I don't think the Python code is setup to deal properly with threads (surprise?). I only see a single `Context` object being created (right when `Raven` is created). Maybe I'm misunderstanding something?

bretthoerner commented:
I didn't `local` instead of `self` on this line: https://github.com/getsentry/raven-python/blob/master/raven/context.py#L33

tkaemming commented on raven/src/main/java/com/getsentry/raven/DefaultRavenFactory.java:
```diff
@@ -75,6 +76,7 @@ public Raven createRavenInstance(Dsn dsn) {
             // https://tomcat.apache.org/tomcat-5.5-doc/servletapi/
             Class.forName("javax.servlet.ServletRequestListener", false, this.getClass().getClassLoader());
             raven.addBuilderHelper(new HttpEventBuilderHelper());
+            raven.addBuilderHelper(new ContextBuilderHelper(raven));
```
I don't think this needs to be in this `try`/`catch` block — shouldn't this be unrelated to the HTTP helper?

tkaemming commented on raven/src/main/java/com/getsentry/raven/RavenContext.java:
```diff
@@ -0,0 +1,117 @@
+public class RavenContext implements AutoCloseable {
+
+    /**
+     * Thread local set of activate context objects. Note that an {@link IdentityHashMap}
```
Is "activate" here supposed to be "active"?

tkaemming commented on raven/src/main/java/com/getsentry/raven/RavenContext.java:
```diff
@@ -0,0 +1,117 @@
+     * is used instead of a Set because there is no identity-set in the Java
+     * standard library.
+     */
+    private static ThreadLocal<IdentityHashMap<RavenContext, RavenContext>> activeContexts =
```
I'm not totally clear on what this is for — is this supposed to facilitate passing a `RavenContext` instance between different threads? If this is the case, it might be a good idea to add something to the `Breadcrumb` object that lets us record the current thread ID and/or name. As far as I know there isn't anything in the "specification" for this yet, but it seems like a thing we'd want to add for platforms where concurrency is going to be more prevalent.

tkaemming commented:
There's decent amount of stuff going on here, and if it's possible, I think it might make sense to break this into two separate patches.

The "low-level" breadcrumb handling (changes to `Breadcrumb`, `BreadcrumbBuilder`, `Event`, `EventBuilder`, and `JsonMarshaller` and associated tests) seems ready to go as-is, but I'm a little apprehensive about moving forward with the thread local `RavenContext` approach before we have a chance to test that with the different logging frameworks. I'm not very familiar with their internals, but I'm worried about interactions if the logging frameworks are doing anything in separate threads (although maybe I'm just being paranoid here?)

bretthoerner commented:
Broke out into two PRs,

https://github.com/getsentry/raven-java/pull/216
https://github.com/getsentry/raven-java/pull/217

Pozo commented:
If you're using `sendEvent(EventBuilder)` rather than `sendEvent(Event)` Raven will not send the previously added Breadcrumbs. It's because `sendEvent(EventBuilder)` will call indirectly `ContextBuilderHelper.helpBuildingEvent` which overrides the event's breadcrumb list.

When you're using default settings this will happen every time since Raven will be created by DefaultRavenFactory which will add ContextBuilderHelper by default.

I've created a test which try to explain what I'm talking about. https://github.com/Pozo/raven-java/commit/b90ee0466714efb108faf5a3d3870b6a15600c58

This is the expected behaviour, or this is a bug ?

bretthoerner commented:
Thanks, that's definitely not intended. I created #314.
